### PR TITLE
Backtest harness core: daily-bar replay of TradingStrategy vs scan signals

### LINKED ISSUE #301: Backtest harness core: daily-bar replay of TradingStrategy vs scan signals
Part of the scanner backtesting epic.

## Ask

Build the daily-bar replay engine:

- Input: scanner_type (+ config parameters), TradingStrategy id, universe, date range.
- Generate historical signals via the scanner's `supports_date_range` path (or query existing ScannerEvents when present).
- Simulate per signal: entry next session open (respecting `entry_type`/`limit_offset_pct`), exit on stop (`stop_pct`), target (`risk_reward_ratio`), or time stop (configurable max hold, default 10 sessions) — conservative intrabar rule: if a daily bar spans both stop and target, count the stop.
- Output per (scanner × strategy): win rate, profit factor, expectancy (R), max drawdown, avg/median hold, trade list.
- Surface as a backend endpoint returning JSON (UI later); persist runs for comparison.

## Acceptance criteria

- Deterministic: same inputs → same stats.
- Handles missing bars/delistings without crashing (skip + count).
- Unit tests cover stop/target/time-stop precedence and the both-sides-hit bar rule.

### PR DESCRIPTION
**Pull request by omniscient**
Closes #301

## Summary
Automated implementation for issue #301.

## Preview
⚠️ _Preview environment failed to start (preview failed: docker compose up --build failed (see preview_build.log tail below)) — endpoint validation was skipped; see the failure comment on the issue._

## Commands
```bash
# Iterate after feedback
docker compose --profile factory run --rm dark-factory "Continue issue #301"

# Tear down preview when done
docker compose --profile factory run --rm dark-factory "Close issue #301"
```

---
*Generated by MarketHawk Dark Factory*

### CODE REVIEW

File: backend/app/services/backtest_service.py [hunk 1]
```diff
@@ -0,0 +1,539 @@
...
+    Optional[Decimal],
+]:
+    """
+    Simulate one trade from an entry bar + ordered subsequent bars.
+
+    Returns: (entry_price, exit_price, exit_reason, hold_sessions, stop_price, target_price)
+
+    Conservative intrabar rule: if bar.low <= stop AND bar.high >= target → stop wins.
+    Time stop: exits at NEXT session open after max_hold_sessions (spec req #6).
+    Limit entry: uses signal_previous_close for limit_price computation (spec §Entry price).
```
Comment (omniscient) [inline backend/app/services/backtest_service.py:142]: **correctness** · **medium** — When `signal_previous_close` is None the limit_price is computed from `open_price` itself, so positive `limit_offset_pct` always fills and the fill semantics diverge from the previous_close-based rule; prefer returning `EXIT_NO_ENTRY` (treat as missing data) over the silent open-based fallback.

File: backend/app/services/backtest_service.py [hunk 2]
```diff
@@ -0,0 +1,539 @@
...
+        db.query(TradingStrategy).filter(TradingStrategy.id == strategy_id).first()
+    )
+    if strategy is None:
+        raise ValueError(f"TradingStrategy id={strategy_id} not found")
+
+    # Snapshot strategy fields at run time for determinism (spec req #9, #11)
+    strategy_snapshot_data = {
+        "entry_type": strategy.entry_type,
+        "stop_pct": str(strategy.stop_pct),
+        "risk_reward_ratio": str(strategy.risk_reward_ratio),
```
Comment (omniscient) [inline backend/app/services/backtest_service.py:372]: **correctness** · **medium** — `universe_as_of=str(date.today())` makes the persisted run record non-deterministic across days for identical inputs, violating "same inputs → same stats"; derive from inputs or the run's created_at, or document it as non-stat provenance.

File: backend/app/services/backtest_service.py [hunk 3]
```diff
@@ -0,0 +1,539 @@
...
+        .distinct()
+        .all()
+    )
+
+    eligible_tickers = list(tickers_with_data)
+
+    result = BacktestResult(
+        run_uuid=str(_uuid.uuid4()),
+        universe_as_of=str(date_type.today()),
+        bars_source="polygon_adjusted",
```
Comment (omniscient) [inline backend/app/services/backtest_service.py:415]: **edge-case** · **medium** — Entry takes the first bar in a next_day..end_date+365d window, so a post-signal halt/data-gap silently enters on a far-future stale bar instead of skip+count per the spec; bound the entry search to the next session (e.g. a few sessions) and otherwise skip/count.

File: backend/app/services/backtest_service.py [hunk 4]
```diff
@@ -0,0 +1,539 @@
...
+
+    wins = [t for t in completed if t.result_r > 0]
+    losses = [t for t in completed if t.result_r <= 0]
+
+    win_rate = len(wins) / len(completed) if completed else None
+
+    gross_profit = sum(t.result_r for t in wins) if wins else 0.0
+    gross_loss = abs(sum(t.result_r for t in losses)) if losses else 0.0
+    profit_factor = (gross_profit / gross_loss) if gross_loss > 0 else None
+
```
Comment (omniscient) [inline backend/app/services/backtest_service.py:230]: **edge-case** · **low** — `losses = [t for t in completed if t.result_r <= 0]` counts break-even (result_r == 0) as a loss, skewing win_rate; use `< 0` for losses and bucket exact zeros separately.

File: backend/app/services/backtest_service.py [hunk 5]
```diff
@@ -0,0 +1,539 @@
...
+    expectancy_r = (
+        sum(t.result_r for t in completed) / len(completed) if completed else None
+    )
+
+    # Max drawdown in R: cumulative sum, track peak then max drop from peak
+    cumulative = 0.0
```
Comment (omniscient) [inline backend/app/services/backtest_service.py:236]: **edge-case** · **low** — `profit_factor = (gross_profit / gross_loss) if gross_loss > 0 else None` returns None for an all-wins run, which consumers may misread as "not computed"; document or use a sentinel for the undefined/infinite case.

File: backend/app/services/backtest_service.py [hunk 6]
```diff
@@ -0,0 +1,539 @@
...
+        .all()
+    )
+
+    found_tickers = {e.ticker for e in db_events}
+    result = [(e.ticker, e.id, e.indicators or {}) for e in db_events]
+
+    missing = [t for t in tickers if t not in found_tickers]
+    if missing:
+        try:
+            descriptor = None
```
Comment (omniscient) [inline backend/app/services/backtest_service.py:303]: **maintainability** · **low** — The nested `try/except Exception: pass` around the `scan_orchestrator._REGISTRY` import swallows all errors and reaches into a private symbol; narrow the except and log, and depend on a public accessor for the registry.

File: backend/app/services/backtest_service.py [hunk 7]
```diff
@@ -0,0 +1,539 @@
...
+                from app.services import scan_orchestrator as _so
+
+                descriptor = _so._REGISTRY.get(scanner_type)
+            except Exception:
+                pass
+
+            if descriptor is not None and descriptor.supports_date_range:
+                loop = asyncio.new_event_loop()
+                try:
+                    raw_signals = loop.run_until_complete(
```
Comment (omniscient) [inline backend/app/services/backtest_service.py]: **maintainability** · **low** — A new `asyncio.new_event_loop()` is created for every missing date inside the day-walk loop and `run_until_complete` raises if invoked under a running loop; create the loop once (or use `asyncio.run`) and guard for an existing loop.

**Review by omniscient — COMMENTED** [review]
## 🏭 Dark Factory Code Review

**0** blocking · **8** advisory finding(s).

### Findings not anchorable to the diff
- [low] edge-case `backend/app/services/backtest_service.py:483` — `stop_dist <= 0` (possible when stop_pct is 0/negative — strategy fields are not validated on this path) silently sets result_r=0.0; validate stop_pct > 0 and log/flag instead of silently zeroing.

File: backend/app/routers/backtest.py [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+
+from app.core.database import get_db
+from app.models.backtest_run import BacktestRun
+from app.models.trading_strategy import TradingStrategy
+from app.schemas.backtest import (
+    BacktestRunDetailResponse,
+    BacktestRunRequest,
+    BacktestRunResponse,
+)
+from app.utils.time import utc_now
```
Comment (omniscient) [inline backend/app/routers/backtest.py:25]: **correctness** · **high** — `created_at=datetime.utcnow()` but `datetime` is not imported (only `uuid`). `create_backtest_run` raises `NameError` on every call. Add `from datetime import datetime` (prefer `app.utils.time.utc_now`).

**Review by omniscient — COMMENTED** [review]
## 🏭 Dark Factory Code Review

**6** blocking · **5** advisory finding(s).

### ⛔ Blocking
- **[critical] correctness** `backend/app/services/backtest_service.py` — `_simulate_trade`/`_get_daily_bars` read `bar.bar_date`, `bar.open_price`, `bar.low_price`, `bar.high_price`, `bar.close_price`, but `StockAggregate` defines `timestamp`, `open`, `high`, `low`, `close`. Every access raises `AttributeError` — the feature cannot run. Use the real column names (and `func.date(timestamp)` or `.timestamp.date()` for the session date).
- **[critical] correctness** `backend/app/services/backtest_service.py` — `_get_daily_bars` queries `StockAggregate` without `timespan == "day"`, so minute bars (the dominant rows) are returned and ordered alongside daily bars, corrupting "entry next session" and intrabar stop/target logic. Add `StockAggregate.timespan == "day"`.
- **[critical] correctness** `backend/app/services/backtest_service.py` — `_run_scanner_for_date` does `scanner_cls = _REGISTRY.get(scanner_type)` then `scanner_cls(config=...)` and `scanner.scan(db=, target_date=, tickers=)`. `_REGISTRY` holds frozen `ScannerDescriptor` dataclasses, not classes; the real interface is `await descriptor.run(tickers, db, event_date, scanner_run=None) -> list[dict]`. The whole `supports_date_range` generation path is broken; call `scan_orchestrator.run(...)` instead.
- **[high] correctness** `backend/app/routers/backtest.py:25` — `created_at=datetime.utcnow()` but `datetime` is not imported (only `uuid`). `create_backtest_run` raises `NameError` on every call. Add `from datetime import datetime` (prefer `app.utils.time.utc_now`).
- **[high] correctness** `backend/app/services/backtest_service.py` — `run_backtest_logic` queries `StockUniverseTicker` but the module never imports it → `NameError`. Add the import.
- **[high] correctness** `backend/app/services/backtest_service.py` — `_run_scanner_for_date` wraps the registry import in `try/except Exception: pass`, then unconditionally references `_REGISTRY`; if the import path is wrong this is a masked `NameError`. Import at module top and let failures surface, or handle the missing-symbol case explicitly.

### Findings not anchorable to the diff
- [critical] correctness `backend/app/services/backtest_service.py` — `_simulate_trade`/`_get_daily_bars` read `bar.bar_date`, `bar.open_price`, `bar.low_price`, `bar.high_price`, `bar.close_price`, but `StockAggregate` defines `timestamp`, `open`, `high`, `low`, `close`. Every access raises `AttributeError` — the feature cannot run. Use the real column names (and `func.date(timestamp)` or `.timestamp.date()` for the session date).
- [critical] correctness `backend/app/services/backtest_service.py` — `_get_daily_bars` queries `StockAggregate` without `timespan == "day"`, so minute bars (the dominant rows) are returned and ordered alongside daily bars, corrupting "entry next session" and intrabar stop/target logic. Add `StockAggregate.timespan == "day"`.
- [critical] correctness `backend/app/services/backtest_service.py` — `_run_scanner_for_date` does `scanner_cls = _REGISTRY.get(scanner_type)` then `scanner_cls(config=...)` and `scanner.scan(db=, target_date=, tickers=)`. `_REGISTRY` holds frozen `ScannerDescriptor` dataclasses, not classes; the real interface is `await descriptor.run(tickers, db, event_date, scanner_run=None) -> list[dict]`. The whole `supports_date_range` generation path is broken; call `scan_orchestrator.run(...)` instead.
- [high] correctness `backend/app/services/backtest_service.py` — `run_backtest_logic` queries `StockUniverseTicker` but the module never imports it → `NameError`. Add the import.
- [high] correctness `backend/app/services/backtest_service.py` — `_run_scanner_for_date` wraps the registry import in `try/except Exception: pass`, then unconditionally references `_REGISTRY`; if the import path is wrong this is a masked `NameError`. Import at module top and let failures surface, or handle the missing-symbol case explicitly.
- [medium] edge-case `backend/app/services/backtest_service.py` — Entry window starts at `signal_date + timedelta(days=1)` assuming the next calendar day is the next session; over weekends/holidays the first available bar can be days out, and `hold_sessions` derived from a calendar span is also affected. Derive the next session from the actual ordered daily-bar list.
- [medium] correctness `backend/app/services/backtest_service.py` — `StockAggregate.open/high/low/close` are `Numeric` → Python `Decimal`. Arithmetic like `entry_price * (1 - strategy.stop_pct / 100)` mixes `Decimal` with `float` and raises `TypeError`. Cast bar prices and strategy params to `float` before math.
- [medium] correctness `backend/app/services/backtest_service.py` — `run.total_signals` is recomputed by re-running `_get_signals_for_date` per day, which ignores scanner-path signals (counts only pre-existing `ScannerEvent`s) and re-issues N queries. Accumulate the count during the main loop.
- [medium] edge-case `backend/app/services/backtest_service.py` — Limit-entry reads `signal_indicators.get("previous_close")`, but `previous_close` is a `ScannerEvent` column, not an `indicators` JSON key, so it almost always falls back to the open price — limit entries effectively behave like market entries. Pull `previous_close` from the event row.
- [low] maintainability `backend/app/services/backtest_service.py` — The `if stop_hit and target_hit:` branch and the `elif stop_hit:` branch have identical bodies (both exit at stop). Collapse to a single `if stop_hit:` to express the conservative both-sides rule without dead duplication.

**Comment by omniscient** [comment]
## Review reconciliation (commit `3068875`)

I verified every finding from both automated review rounds against the reviewed HEAD (`08a39cd`).

**The second review round (6 blocking · 5 advisory) is stale** — its blocking findings describe the *original* implementation and were all already fixed by `517d4eb` ("address 8 code review findings"), which is an ancestor of the reviewed commit:

| Reported "blocking" | Actual code at `08a39cd` |
|---|---|
| `bar.bar_date`/`bar.open_price` → AttributeError | uses `bar.open`/`bar.low`/`bar.timestamp` |
| `_get_daily_bars` missing `timespan == "day"` | present |
| `_REGISTRY.get()` / `scanner_cls(...)` | uses `scan_orchestrator.get_all()` + `descriptor.run()` |
| router `datetime.utcnow()` NameError | `utc_now()` (imported) |
| `StockUniverseTicker` not imported | imported |
| masked `NameError` in `try/except` | handles `ImportError` explicitly |

The 8 first-review inline comments (break-even-as-loss, `date.today()` non-determinism, `+365d` stale-entry window, event-loop-per-date, etc.) were likewise already addressed.

**Two findings were genuinely still live and are fixed in `3068875`:**

1. **[medium] limit-entry `previous_close` never sourced** — `previous_close` is a `ScannerEvent` *column*, not an `indicators` JSON key, so `indicators.get("previous_close")` was always `None` for DB-sourced signals. Combined with the earlier change to return `EXIT_NO_ENTRY` when `previous_close is None`, every DB-sourced limit-entry signal silently produced **zero trades**. `_get_signals_for_date` now folds the column into the indicators dict; added a regression test (verified failing on pre-fix code).
2. **[low] maintainability** — collapsed the duplicate `if stop_hit and target_hit:` / `if stop_hit:` branches into a single `if stop_hit:` (behavior identical; covered by `test_intrabar_both_sides_hit_stop_wins`).

All 22 service tests pass; ruff lint + format clean.
